### PR TITLE
Ruby Capstone: Catalog of My Things

### PR DESCRIPTION
**Pull request by AliRezaBmeDu**
In this branch, we have met the following requirements of the capstone project- **Catalog of My Things**
The project is described in this [short video](https://drive.google.com/file/d/1ZQ0hriUhin8m1vRhHBsAbc4W2xjC07Ye/view?usp=sharing) [This is the new "Below 5 min version" as per the first reviewer's requirement]

### General requirements

- Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).

- Make sure to use the correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
- Make sure to document the work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

### Logic
- Each class from the [UML class diagram](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/images/catalog_of_my_things.png) should be created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.
- All properties visible in the diagram should be defined and set up in the constructor method. Exception: properties for the 1-to-many relationships should NOT be set in the constructor method. Instead, they should have a custom setter method created.
- All methods visible in the diagram should be implemented:
- add_item method (in all classes that have that method)

  1.  should take an instance of the Item class as an input.
  2.  should add the input item to the collection of items.
  3.  should add self as a property of the item object (by using the correct setter from the item object).

- can_be_archived?() in the Item class

  1.  should return true if published_date is older than 10 years.
  2.  otherwise, it should return false.

- move_to_archive() in the Item class

    1. should reuse can_be_archived?() method.
    2. should change the archived property to true if the result of the can_be_archived?() method is true.
    3. should do nothing if the result of the can_be_archived?() method is false.

- can_be_archived?() in the Book class

    1. should override the method from the parent class.
    2. should return true if parent's method returns true OR if cover_state equals to "bad".
    3. otherwise, it should return false.

- can_be_archived?() in the MusicAlbum class

  1.   should override the method from the parent class.
  2.   should return true if parent's method returns true AND if on_spotify equals true.
  3.   otherwise, it should return false.

- can_be_archived?() in the Game class

  1.   should override the method from the parent class.
  2.   should return true if parent's method returns true AND if last_played_at is older than 2 years.
  3.   otherwise, it should return false.

- Add unit tests for all implemented methods.
- Create a main.rb file that will serve as the console app entry-point.
   The console app, at the start, should:

    1. Present the user with a list of options to perform.
    2. Let users choose an option.
    3. If needed, ask for parameters for the option.
    4. Have a way to quit the app.
    5. The following options should be available:
    6. List all books
    7. List all music albums
    9. List of games
    10. List all genres (e.g 'Comedy', 'Thriller')
    11. List all labels (e.g. 'Gift', 'New')
    12. List all authors (e.g. 'Stephen King')
    13. List all sources (e.g. 'From a friend', 'Online shop')
    14. Add a book
    15. Add a music album
    16. Add a game
    18. All data should be preserved by saving collections in .json files.

### Database
- Create a schema.sql file with tables that will be analogical to the structure of the classes in the app. As you cannot implement inheritance in the database tables - add all properties and associations from the parent Item class as table columns to all tables based on the child classes. Thanks to that it can be used to store data in the future.

### CODE REVIEW

File: spec/book_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+      book = Book.new('First publisher', 'bad', '17-10-2023')
+      expect(book.publisher).to eq('First publisher')
+    end
+
+    it 'should have bad as the cover_state' do
+      book = Book.new('First publisher', 'bad', '17-10-2023')
+      expect(book.cover_state).to eq('bad')
+    end
+  end
+end
```
Comment (ghost) [inline spec/book_spec.rb]: - [x] Make sure to test all the methods within your classes. When it comes to your book class, don't forget to test the "can_be_archived?" method! 😊

File: item.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    @author = nil
+    @source = nil
+    @label = nil
+  end
+
+  def add_label(book_label)
+    @label = book_label
+  end
+
+  def add_genre(album_genre)
```
Comment (ghost) [inline item.rb:20]: - [x] I've also noticed that your item class hasn't been tested. Please, don't forget to give it a test run! 🧪

**Review by frank1738 — CHANGES_REQUESTED** [review]
Hi Team, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



YesYouAreGIF (2)

## Highlights

✅ Your UI is impressive and super user-friendly!  👍 
✅ You've nailed the error handling in those tricky edge case scenarios! 🛡️
✅ Kudos for testing most of your classes! 🧪
✅ Your code is clean and tidy! 👌🧹👏


## Required Changes ♻️

- [x] 🌟 Awesome job on your video presentation! 📽️👏 However, it's super important to keep in mind that the project guidelines specify a max length of 5 minutes for the video, while yours is 8 minutes long. 🕒😅 I'd strongly recommend recording a shorter video to ensure it aligns with the project requirements. 📌✂️ Thanks!
_Check the comments under the review._

## Optional suggestions

N/A
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by DeliceLydia — APPROVED** [review]
## Hi @Team, 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


### Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**